[threads] Switch foreign threads to GC Safe in mono_thread_detach

monojenkins opened the pull request:
!! This PR is a copy of mono/mono#20435,  please do not edit or review it in this repo !!<br/>Do not automatically approve this PR:<br/><br/>* Consider how the changes affect configurations in this repo,<br/>* Check effects on files that are not mirrored,<br/>* Identify test cases that may be needed in this repo.<br/><br/>!! Merge the PR only after the original PR is merged !!<br/><hr/><br/>
Addresses https://github.com/mono/mono/issues/20290 and https://github.com/mono/mono/issues/20283

If a foreign thread (that was created outside the runtime) calls
`mono_thread_detach`, leave it in a preemptively-suspendable state, since we can't expect it to coop suspend.

Conversely in `mono_thread_attach` (external only), ensure that we always leave
the thread in GC Unsafe (aka RUNNING) state, for cases like

    while (cond) {
      t = mono_thread_attach (domain);
      <...>
      mono_thread_detach (t);
    }

---

To make this work, we also make `mono_thread_detach` and `mono_thread_attach` external only.  The runtime should use `mono_thread_internal_detach` and `mono_thread_internal_attach`.

lambdageek commented:
@lateralusX I pushed https://github.com/dotnet/runtime/pull/42758/commits/f6e9c0182b5252a98f4a8256b9aea691d83a2aeb into this PR to add a call in EventPipe to `mono_thread_internal_attach` (MONO_PROFILER_API) instead of `mono_thread_attach` (MONO_API) - I think this means I need to fix the non-static EP build by adding an implementation using the function table, right?

lateralusX commented:
@lambdageek don't worry about that at the moment, as part of https://github.com/dotnet/runtime/pull/41872 I pushed these methods into support methods that are already hooked up to the function table, so depending on what PR that get merged first we just need to adjust call within that function, eventpipe_thread_attach.

lambdageek committed:
[threads] Switch foreign threads to GC Safe in mono_thread_detach

Addresses https://github.com/mono/mono/issues/20290 and https://github.com/mono/mono/issues/20283

If a foreign thread (that was created outside the runtime) calls
`mono_thread_detach`, leave it in a preemptively-suspendable state, since we can't expect it to coop suspend.

Conversely in `mono_thread_attach` (external only), ensure that we always leave
the thread in GC Unsafe (aka RUNNING) state, for cases like

    while (cond) {
      t = mono_thread_attach (domain);
      <...>
      mono_thread_detach (t);
    }

---

To make this work, we also make `mono_thread_detach` and `mono_thread_attach` external only.  The runtime should use `mono_thread_internal_detach` and `mono_thread_internal_attach`.

lambdageek committed:
fixup eventpipe to use mono_thread_internal_{attach,detach}

lambdageek commented:
The (Libraries Test Run release mono Linux x64 Debug) failure in SetDateTimeMax is due to https://github.com/dotnet/runtime/issues/43166
